src/Npgsql/Replication/Logical/Protocol/NpgsqlLogicalReplicationConnectionExtensions.cs

```diff
@@ -0,0 +1,352 @@
+        /// knows which WAL segments are still needed by the standby.
+        /// </param>
+        /// <param name="walLocation">The WAL location to begin streaming at.</param>
+        /// <param name="options">The collection of options passed to the slot's logical decoding plugin.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.
+        /// The default value is <see cref="CancellationToken.None"/>.</param>
+        /// <returns>A <see cref="Task{T}"/> representing an <see cref="IAsyncEnumerable{T}"/> that
+        /// can be used to stream WAL entries in form of <see cref="LogicalReplicationProtocolMessage"/> instances.</returns>
+        [PublicAPI]
+        public static Task<IAsyncEnumerable<LogicalReplicationProtocolMessage>> StartReplication(
+            this NpgsqlLogicalReplicationConnection connection, NpgsqlPgOutputReplicationSlot slot,
+            NpgsqlPgOutputPluginOptions options, NpgsqlLogSequenceNumber? walLocation = null, CancellationToken cancellationToken = default)
+        {
+            using (NoSynchronizationContextScope.Enter())
+                return StartReplicationInternal(connection, slot, walLocation, options, cancellationToken);
+        }
+
+        static async Task<IAsyncEnumerable<LogicalReplicationProtocolMessage>> StartReplicationInternal(
+            NpgsqlLogicalReplicationConnection connection, NpgsqlPgOutputReplicationSlot slot,
+            NpgsqlLogSequenceNumber? walLocation, NpgsqlPgOutputPluginOptions options, CancellationToken cancellationToken = default)
+        {
+            var stream = await connection.StartReplicationInternal(commandBuilder =>
+            {
+                commandBuilder.Append("SLOT ").Append(slot.SlotName).Append(' ');
+                Internal.NpgsqlLogicalReplicationConnectionExtensions.AppendCommon(commandBuilder, walLocation, options.GetOptionPairs(), slot.ConsistentPoint);
+            }, true, cancellationToken);
+#pragma warning disable CA2016
+            // ReSharper disable once MethodSupportsCancellation
+            return StreamData(stream, connection);
+#pragma warning restore CA2016
+        }
+
+        static async IAsyncEnumerable<LogicalReplicationProtocolMessage> StreamData(IAsyncEnumerable<NpgsqlXLogDataMessage> messageStream, NpgsqlLogicalReplicationConnection connection, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+        {
+            // Hack: NAMEDATALEN is a constant in PostgreSQL which can be changed at compile time. It's probably saner to query max_identifier_length on connection startup.
+            // See https://www.postgresql.org/docs/current/runtime-config-preset.html
+            const int NAMEDATALEN = 64;
+            var buf = connection.Connector!.ReadBuffer;
+            await foreach (var xLogData in messageStream.WithCancellation(cancellationToken))
```

Comment:
Ah yes, I remember this lovely hack :rofl: where we get an XLogData message, but just go directly into the buffer which is at the start of the data... Ugly but works :)